modules/configuration/src/main/java/org/apache/ignite/configuration/ConfigurationChanger.java

```diff
@@ -0,0 +1,133 @@
+
+        // TODO: iterate over data and fill Configurators
+    }
+
+    public void registerConfiguration(RootKey<?> key, Configurator<?> configurator) {
+        registry.put(key, configurator);
+    }
+
+    public <T extends ConfigurationTree<?, ?>> void change(Map<RootKey<?>, TraversableTreeNode> changes) {
+        Map<String, Serializable> allChanges = changes.entrySet().stream()
+            .map((Map.Entry<RootKey<?>, TraversableTreeNode> change) -> convertChangesToMap(change.getKey(), change.getValue()))
+            .flatMap(map -> map.entrySet().stream())
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+        boolean success = false;
+
+        List<ValidationIssue> validationIssues = Collections.emptyList();
+
+        while (!success) {
+            validationIssues = validate(changes);
+
+            final int version = this.version.get();
+
+            if (validationIssues.isEmpty())
+                success = configurationStorage.write(allChanges, version);
+            else
+                break;
+        }
+
+        if (!validationIssues.isEmpty())
+            throw new ConfigurationValidationException(validationIssues);
+    }
+
+    private List<ValidationIssue> validate(Map<RootKey<?>, TraversableTreeNode> changes) {
+        List<ValidationIssue> issues = new ArrayList<>();
+
+        for (Map.Entry<RootKey<?>, TraversableTreeNode> entry : changes.entrySet()) {
+            RootKey<?> rootKey = entry.getKey();
+            TraversableTreeNode changesForRoot = entry.getValue();
+
+            final Configurator<?> configurator = registry.get(rootKey);
+
+            List<ValidationIssue> list = configurator.validateChanges(changesForRoot);
+            issues.addAll(list);
+        }
+
+        return issues;
+    }
+
+    private Map<String, Serializable> convertChangesToMap(RootKey<?> rootKey, TraversableTreeNode node) {
+        Map<String, Serializable> values = new HashMap<>();
+
+        node.accept(null, new ConfigurationVisitor() {
+
+            String currentKey = rootKey.key();
+
+            @Override public void visitLeafNode(String key, Serializable val) {
+                values.put(currentKey + "." + key, val);
+            }
+
+            @Override public void visitInnerNode(String key, InnerNode node) {
+                String previousKey = currentKey;
+
+                if (key != null)
+                    currentKey += "." + key;
+
+                node.traverseChildren(this);
+
+                currentKey = previousKey;
+            }
+
+            @Override public <N extends InnerNode> void visitNamedListNode(String key, NamedListNode<N> node) {
+                String previousKey = currentKey;
+
+                if (key != null)
+                    currentKey += "." + key;
+
+                for (String namedListKey : node.namedListKeys()) {
+                    String loopPreviousKey = currentKey;
+                    currentKey += "." + namedListKey;
```

Comment:
Keep in mind that keys in named lists must be escaped. Methods for that are locted in this PR right now: https://github.com/apache/ignite-3/pull/40/files